File: solutions/ruby/raindrops/1/raindrops.rb
class Raindrops

  DROPS = {
    3 => 'Pling',
    5 => 'Plang',
    7 => 'Plong'
  }
  private_constant :DROPS

  def self.convert number
    new(number).sound
  end

  private

  attr_accessor :number

  def initialize number
    self.number = number
  end

  def generate_sound
    DROPS.select { |factor, _| number % factor == 0 }
      .values
      .inject(:+) or number.to_s
  end

  public

  def sound
    # memoized
    sound ||= generate_sound
  end

end

File: solutions/haskell/raindrops/1/src/Raindrops.hs
module Raindrops (convert) where

convert :: Int -> String
convert n
  | 3 `divides` n && 5 `divides` n && 7 `divides` n = "PlingPlangPlong"
  | 3 `divides` n && 7 `divides` n = "PlingPlong"
  | 3 `divides` n && 5 `divides` n = "PlingPlang"
  | 5 `divides` n && 7 `divides` n = "PlangPlong"
  | 3 `divides` n = "Pling"
  | 5 `divides` n = "Plang"
  | 7 `divides` n = "Plong"
  | otherwise = show n
  
divides :: Int -> Int -> Bool
divides m n = mod n m == 0
